File: micro.php
<?php

ini_set('memory_limit','1G');

function f1() {
	$start = microtime(true);
	$x = 10;
	for ($i = 0; $i < 100000; ++$i) {
		$x = ( 2 * $x ) + $x + 1;
	}
#	print "t1 answer %d time %f" % (x, float(time.clock()) - start) 
	printf( "f1 result %d time %f seconds\n" , $x, microtime(true) - $start );
}

function f1a() {
	$start = microtime(true);
	$x = 10;
	$i = 0;
	do {
		$x = ( 2 * $x ) + $x + 1;
		++$i;
	} while ($i < 100000);
#	print "t1 answer %d time %f" % (x, float(time.clock()) - start) 
	printf( "f1a (different loop) result %d time %f seconds\n" , $x, microtime(true) - $start );
}

function f2() {
	$t0 = microtime(true);

	$nStrings = 1000;
	$nCats = 100000;

	$a1 = array();
	$c = ord( "a" );
	$a1[0] = (string) chr($c);

	for ($x = 1; $x < $nStrings; $x++) {
		$a1[ $x ] = $a1[ $x - 1] . chr($c);
	}

	$t1 = microtime(true);

	$a2 = array();
	for ($x = 0; $x < $nCats; $x++) {
		$a2[ $x ] = $a1[ $x % count($a1) ] . 
					$a1[ ( $x + 7 ) % count($a1) ] . 
					$a1[ ( $x + 17 ) % count($a1) ] . 
					$a1[ ( $x + 27 ) % count($a1) ] . 
					$a1[ ( $x + 37 ) % count($a1) ] . 
					$a1[ ( $x + 47 ) % count($a1) ] .
					$a1[ ( $x + 57 ) % count($a1) ] ;
	}

	$t2 = microtime(true);

	printf( "f2 string create %f sec cat with dot %f\n" , $t1 - $t0, $t2 - $t1 );

}


function f5() {
  $start = microtime(true);
  for ($i = 0; $i < 1000000; ++$i) {
     $a = array();
     for ($a1 = 0; $a1 < 50; ++$a1 ) {
        $a[$a1] = strval($a1);
     }
  }
  printf( "t5 %f seconds\n",microtime(true) - $start) ;
}

print "quick performance test\n";
f1();
f1a();
f2();
f5();
print "end quick performance test\n";
?>

File: micro.cpp
#include <iostream>
#include <stdio.h>
#include <chrono>
#include <vector>
#include <map>

using namespace std;

void f1()
{
	chrono::steady_clock::time_point start = chrono::steady_clock::now();
	int x = 10;
	for (int i = 0; i < 100000; i++) {
		x = ( 2 * x ) + x + 1;
  }

  chrono::steady_clock::time_point end = chrono::steady_clock::now();

  std::cout << "f1 took " 
            << chrono::duration_cast<chrono::milliseconds>(end - start).count() 
            << " milliseconds. " << std::endl;
}

void f2()
{
  chrono::steady_clock::time_point t0 = chrono::steady_clock::now();

  int nStrings = 1000;
  int nCats = 100000;

  std::vector<std::string> a1(nStrings);
  char c = 'a';
  std::string myStr(&c);
  a1[0] = myStr;

  for (int i=1;i<nStrings;i++) {
    if (c == 0) c++;
    a1[i] = a1[i-1] + c;
    c++;
  }

  // Do the vector work with push, like many languages do

  chrono::steady_clock::time_point t1 = chrono::steady_clock::now();

  vector<string> a2;
  for (int i = 0; i < nCats ; i++) {
    a2.push_back( a1[i % a1.size()] + a1[(i+7)%a1.size()] + a1[(i+17)%a1.size()] + 
      a1[(i+27)%a1.size()] + a1[(i+37)%a1.size()] + a1[(i+47)%a1.size()] + a1[(i+57)%a1.size()] );
  }

  chrono::steady_clock::time_point t2 = chrono::steady_clock::now();

  // do the vector work with a preallocated array, as is the common thing in C
  vector<string> a3(nCats);
  for (int i = 0; i < nCats ; i++) {
    a3[i] = a1[i % a1.size()] + a1[(i+7)%a1.size()] + a1[(i+17)%a1.size()] + 
      a1[(i+27)%a1.size()] + a1[(i+37)%a1.size()] + a1[(i+47)%a1.size()] + a1[(i+57)%a1.size()] ;
  }

  chrono::steady_clock::time_point t3 = chrono::steady_clock::now();

  std::cout << "f2 createstr only " 
            << chrono::duration_cast<chrono::milliseconds>(t1 - t0).count() 
            << " millisec , plus catenating " 
            << chrono::duration_cast<chrono::milliseconds>(t2 - t1).count()
            << " millisec, optimzied vectors "
            << chrono::duration_cast<chrono::milliseconds>(t3 - t2).count()
            << std::endl ; 
}

void f5()
{
  chrono::steady_clock::time_point t0 = chrono::steady_clock::now();

  // the basic and most idomatic way to fulfill this
  for (int i = 0; i < 1000000;i++) {
    map<int, string> a;
    for (int j = 0; j < 50; j++) {
      a[j] = to_string(j);
    }
  }
  // NB: don't have to delete things. These are on the stack?

  chrono::steady_clock::time_point t1 = chrono::steady_clock::now();

  // try with just one, and clearing it
  map<int, string> a1;
  for (int i = 0; i < 1000000;i++) {
    for (int j = 0; j < 50; j++) {
      a1[j] = to_string(j);
    }
    a1.clear();
  }

  chrono::steady_clock::time_point t2 = chrono::steady_clock::now();

  // Re-use the same string --- is it the string create which is long?
  map<int, const char *> a2;
  for (int i = 0; i < 1000000;i++) {
    for (int j = 0; j < 50; j++) {
      a2[j] = "123";
    }
    a2.clear();
  }

  chrono::steady_clock::time_point t3 = chrono::steady_clock::now();


  std::cout << "f5 " 
            << chrono::duration_cast<chrono::milliseconds>(t1 - t0).count() 
            << " ms standard, "
            << chrono::duration_cast<chrono::milliseconds>(t2 - t1).count() 
            << " ms reusing map, "
            << chrono::duration_cast<chrono::milliseconds>(t2 - t1).count() 
            << " ms constant string "
            << std::endl ;

}

int main(){

  printf( "quick performance test\n" );
  f1();
  f2();
  f5();
  std::cout << "end quick performance test\n";
}